### ISSUE TITLE
FEATURE: Validation League

### ISSUE THREAD

**Issue by JordanKotsop** [JordanKotsop/context-dao#7]
## Feature: Validation League

**Part of Epic:** #1
**Priority:** Nice-to-have | **Size:** L | **Sprint:** 3

---

## Why This Feature Exists

This is the Bittensor-inspired incentive loop. Skill owners post bounties, prompt optimizers submit improved versions, and a Referee Agent scores them against a validation set. If the score improves, x402 payment is automatically triggered to the optimizer. This creates a flywheel: better skills → more usage → more revenue → bigger bounties → better optimizers. It's a "Mini-Subnet" for prompt engineering.

---

## User Stories

**As a skill owner,**
I want to post a bounty ("Pay $X for every 1% accuracy increase or 10% token reduction"),
so that the crowd optimizes my cognitive assets for me.

**As a prompt optimizer,**
I want to download a skill's current version and validation set, submit my improved version, and get paid automatically if my score beats the current best,
so that I earn money from my prompt engineering expertise.

**As the platform,**
I want an automated, trustless validation pipeline,
so that optimization payouts are objective and don't require manual review.

---

## Acceptance Criteria

- [ ] **Bounty creation** — Skill owners can create a bounty:
  - Skill ID to optimize
  - Reward per % accuracy increase (USDC via x402)
  - Reward per % token reduction
  - Max total bounty pool
  - Expiration date
- [ ] **Submission flow** — Optimizers submit:
  - New skill `.md` file (v2)
  - Which bounty they're targeting
  - Their wallet address for payout
- [ ] **Referee Agent** — Automated validation:
  - Loads the validation set (from Feature #2 companion `.json`)
  - Runs both v1 and v2 against all test questions
  - Scores using configured match type (keyword, LLM-as-Judge, exact)
  - Calculates marginal improvement (accuracy delta, token delta)
  - Signs a verification message with the results
- [ ] **Auto-payout** — If improvement confirmed:
  - Calculate reward: `delta_accuracy * reward_per_percent`
  - Trigger x402 payment from bounty pool to optimizer wallet
  - Update skill to new version (with owner approval gate)
- [ ] **Leaderboard** — Public ranking of:
  - Top optimizers (by total earnings)
  - Most improved skills (by accuracy gain)
  - Active bounties (sorted by reward size)
- [ ] **Anti-gaming:**
  - Validation set is NOT included in the skill (stored separately)
  - Referee runs in isolated environment
  - Overfitting detection: test on held-out set not shown to optimizers

---

## Technical Notes

**Validation Pipeline:**
```
Optimizer submits v2.md
    ↓
[Referee Agent]
    ├── Load validation.json (Q&A pairs)
    ├── Run v1.md against all questions → score_v1
    ├── Run v2.md against all questions → score_v2
    ├── Calculate delta = score_v2 - score_v1
    ├── Calculate token_delta = tokens_v1 - tokens_v2
    ↓
If delta > 0:
    ├── Sign verification message
    ├── Trigger x402 payout (delta * reward_rate)
    └── Queue v2 for owner approval
Else:
    └── Return "No improvement detected"
```

**Scoring Methods:**
1. **Keyword match:** Expected output contains required terms (fast, deterministic)
2. **LLM-as-Judge:** Claude Haiku rates response quality 1-10 (nuanced, slower)
3. **Exact match:** For factual/numerical answers (strictest)

**Stack:**
- Python validation engine
- Runs as a background job / queue (not real-time)
- Uses same LLM client as Blind Inference Runner

**Key Files:**
```
validation/
├── referee.py           # Core validation logic
├── scoring/
│   ├── keyword.py       # Keyword match scorer
│   ├── llm_judge.py     # LLM-as-Judge scorer
│   └── exact.py         # Exact match scorer
├── bounty/
│   ├── models.py        # Bounty data model
│   └── payout.py        # x402 payout trigger
└── api/
    ├── bounties.py      # Bounty CRUD endpoints
    └── submissions.py   # Optimization submission endpoints
```

---

## Dependencies

- **Blocked by:** #2 (Skill Asset Standard — needs validation sets), #4 (x402 Gate — payout mechanism), #5 (Blind Inference — uses same LLM infrastructure)
- **Blocks:** Nothing — this is the incentive layer on top

---

## Bittensor Parallel

| Bittensor Concept | ContextDAO Equivalent |
|---|---|
| Subnet | Skill category (e.g., "Finance/Tax") |
| Miner | Prompt optimizer submitting v2 |
| Validator | Referee Agent running validation set |
| Loss function | Accuracy delta + token reduction |
| TAO reward | USDC payout via x402 |
| Weight setting | Owner approval + auto-version bump |

---

*Created with `/plan-features` — Product Owner enriched*
